### PR TITLE
[relay][topi] Add operation relay.nn.dilate() which calls topi.nn.dilate()

### PR DESCRIPTION
**Pull request by notoraptor**
Hi! I made this PR to add operation `dilate` into relay.

My main goal is to be able to implement a complete version of `conv2d_transpose` in relay which supports all values for groups, dilations, output padding and strides (code based on Theano mplementation: https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/abstract_conv.py#L2927 ). I needed operation `dilation` to do that:

```python
from tvm import relay


def conv2d_transpose(
    input_shape, kern_shape, strides, padding, dilation, output_padding, groups
):
    h_out = (
        (input_shape[2] - 1) * strides[0]
        - 2 * padding[0]
        + dilation[0] * (kern_shape[2] - 1)
        + output_padding[0]
        + 1
    )
    w_out = (
        (input_shape[3] - 1) * strides[1]
        - 2 * padding[1]
        + dilation[1] * (kern_shape[3] - 1)
        + output_padding[1]
        + 1
    )

    data = relay.var("data", shape=input_shape)
    weight = relay.var("weight", shape=kern_shape)
    data_dilated = relay.nn.dilate(data, (1, 1) + strides)
    data_padded = relay.nn.pad(
        data_dilated,
        ((0, 0), (0, 0), (0, output_padding[0]), (0, output_padding[1]),),
    )

    # Pre-process kernel,
    # from (m0, m1, m2, m3) to (m1 * g, m0 // g, m2, m3).
    mshp0 = kern_shape[0] // groups
    c_out = kern_shape[1] * groups
    kern = relay.reshape(weight, (groups, mshp0) + kern_shape[1:])
    # => (g, m0 // g, m1, m2, m3)
    kern = relay.op.transpose(kern, axes=(1, 0, 2, 3, 4))
    # => (m0 // g, g, m1, m2, m3)
    kern = relay.reshape(kern, (mshp0, c_out, kern_shape[-2], kern_shape[-1]))
    # => (m0 // g, m1 * g, m2, m3)
    kern = relay.op.transpose(kern, (1, 0, 2, 3))
    # => (m1 * g, m0 // g, m2, m3)
    # Kernel 2 latest dimensions must be flipped
    kern = relay.op.transform.reverse(kern, 2)
    kern = relay.op.transform.reverse(kern, 3)
    # End pre-processing kernel.

    img = relay.nn.conv2d(
        data_padded,
        kern,
        groups=groups,
        channels=c_out,
        padding=[(kern_shape[2 + i] - 1) * dilation[i] for i in range(2)],
        dilation=dilation,
    )

    if any(p != 0 for p in padding):
        img = relay.op.transform.strided_slice(
            data=img,
            begin=[0, 0, padding[0], padding[1]],
            end=[None, None, h_out + padding[0], w_out + padding[1],],
        )

    f = relay.Function([data, weight], img)
    return f

```

### CODE REVIEW

**Comment by tqchen** [comment]
cc @vinx13 @icemelon9 @kevinthesun

File: src/relay/op/nn/nn.cc [hunk 1]
```diff
@@ -961,6 +961,56 @@ Do log on the data - do not accept logits.
...
 
 
+/////
+// relay.nn.dilate
+TVM_REGISTER_NODE_TYPE(DilateAttrs);
+
+bool DilateRel(const Array<Type>& types,
+                    int num_inputs,
+                    const Attrs& attrs,
+                    const TypeReporter& reporter) {
```
Comment (siju-samuel) [inline src/relay/op/nn/nn.cc:971]: Align the arguments

File: src/relay/op/nn/nn.cc [hunk 2]
```diff
@@ -961,6 +961,56 @@ Do log on the data - do not accept logits.
 .add_type_rel("CrossEntropy", CrossEntropyRel);
 
 
+/////
```
Comment (siju-samuel) [inline src/relay/op/nn/nn.cc:964]: remove ////, down also

File: include/tvm/relay/attrs/nn.h [hunk 3]
```diff
@@ -442,6 +442,18 @@ struct Conv2DTransposeAttrs : public tvm::AttrsNode<Conv2DTransposeAttrs> {
   }
 };
 
+////
```
Comment (siju-samuel) [inline include/tvm/relay/attrs/nn.h:445]: remove ////, down also

File: python/tvm/relay/op/nn/nn.py [hunk 4]
```diff
@@ -1347,6 +1347,25 @@ def pad(data,
...
 
+def dilate(data, strides):
+    """Dilate data with zeros.
+
+    Parameters
+    ----------
+    data : tvm.relay.Expr
+        n-D, can be any layout.
+
+    strides : <tuple of n <int>
```
Comment (siju-samuel) [inline python/tvm/relay/op/nn/nn.py:1358]: `tuple of <int>`

**Comment by notoraptor** [comment]
@siju-samuel fixed!

**Comment by tqchen** [comment]
cc @Huyuwei @kazum please help manage this PR

File: python/tvm/relay/op/nn/_nn.py [hunk 5]
```diff
@@ -458,6 +458,15 @@ def compute_cross_entropy(attrs, inputs, out_dtype):
 reg.register_pattern("nn.cross_entropy", OpPattern.OPAQUE)
 
 
+# dilate
+@reg.register_compute("nn.dilate")
+def compute_dilate(attrs, inputs, out_dtype):
+    return [topi.nn.dilate(inputs[0], attrs.strides)]
+
+reg.register_broadcast_schedule("nn.dilate")
+reg.register_pattern("nn.dilate", OpPattern.OPAQUE)
```
Comment (kazum) [inline python/tvm/relay/op/nn/_nn.py:467]: I wonder if the op pattern is better to be injective.

Comment (notoraptor) [reply python/tvm/relay/op/nn/_nn.py]: @kazum Done!

**Comment by kazum** [comment]
Thanks @notoraptor @siju-samuel !
